src/main/java/org/tikv/txn/gc/GCWorker.java

```diff
@@ -0,0 +1,492 @@
+    // acquire lock
+    boolean ok = pdClient.lock(GC_LEADER_UUID_KEY, l_uuid) && doCheckLeader();
+    // release lock
+    pdClient.unlock(GC_LEADER_UUID_KEY);
+    return ok;
+  }
+
+  private boolean doCheckLeader() {
+    long leader = getLong(GC_LEADER_UUID_KEY);
+    logger.debug(String.format("[gc worker] got leader: %s", Long.toHexString(leader)));
+    if (leader == l_uuid) {
+      try {
+        putLong(GC_LEADER_LEASE_KEY, getNow() + gcWorkerLease);
+      } catch (Exception e) {
+        logger.warn("[gc worker] gc leader lease key update fails.", e);
+        return false;
+      }
+      return true;
+    }
+    long lease = getLong(GC_LEADER_LEASE_KEY);
+    long now = getNow();
+    if (lease == 0 || lease < now) {
+      logger.debug(String.format("[gc worker] register %s as leader", uuid));
+      try {
+        pdClient.txn().Then(
+            toPutOp(GC_LEADER_UUID_KEY, uuid),
+            toPutOp(GC_LEADER_LEASE_KEY, now + gcWorkerLease)
+        ).commit().get(500, TimeUnit.MILLISECONDS);
+      } catch (Exception e) {
+        logger.warn("[gc worker] gc leader lease key init fails.", e);
+        return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  private void runGCJob(long safePoint) {
+    try {
+      resolveLocks(safePoint);
+    } catch (Exception e) {
+      logger.error(String.format("[gc worker] %s resolve locks returns an error %s", uuid, e));
+      return;
+    }
+    doGC(safePoint);
+  }
+
+  private void doGC(long safePoint) {
+    long concurrency = getLongOrElse(GC_CONCURRENCY_KEY, gcDefaultConcurrency);
+    if (concurrency < gcMinConcurrency) {
+      concurrency = gcMinConcurrency;
+    }
+    if (concurrency > gcMaxConcurrency) {
+      concurrency = gcMaxConcurrency;
+    }
+
+    doGCInternal(safePoint, concurrency);
+  }
+
+  private void doGCInternal(long safePoint, long concurrency) {
+    putLong(GC_SAFE_POINT_KEY, safePoint);
+    logger.info(String.format("[gc worker] %s start gc, concurrency %d, safePoint: %d.", uuid, concurrency, safePoint));
+
+    try {
+      gcTaskThreadPool = Executors.newFixedThreadPool((int) concurrency);
+      gcTaskService = new ExecutorCompletionService<>(gcTaskThreadPool);
+      for (int i = 0; i < concurrency; i++) {
+        new Thread(() -> newGCTaskWorker(uuid).run()).start();
+      }
+      ByteString key = ByteString.EMPTY;
+      while (true) {
+        BackOffer backOffer = ConcreteBackOffer.newCustomBackOff(BackOffer.GcOneRegionMaxBackoff);
+        GCTask task = genNextGCTask(backOffer, safePoint, key);
+        gcTaskService.submit(() -> task);
+        key = task.endKey;
+        if (key.equals(ByteString.EMPTY)) {
+          awaitGCComplete();
+          return;
+        }
+      }
+    } catch (Exception e) {
+      awaitGCComplete();
+    }
+  }
+
+  private void awaitGCComplete() {
+    try {
+      if (gcTaskThreadPool != null) {
+        gcTaskThreadPool.awaitTermination(gcWorkerLease, TimeUnit.MILLISECONDS);
```

Comment:
Is the parameter the timeout? If so, this doesn't seems to be correct. It might take long time to finish GC, while the tick should continue. The leader lease should still be updated in leaderTick.

Reply:
Found out that I forgot to make runGCJob async. My bad.

Reply:
done.

Reply:
Emmm, so is the parameters the timeout? It doesn't seems to be correct since it may take very long time to finish.

Reply:
There is no other way to identify whether GC is completed or the task executor was lost.

Reply:
Maybe you can add a callback to each GCTask which is invoked on finish, and wait for it to be called enough times?